ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.server;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// reference:
+// https://gist.github.com/danielflower/f54c2fe42d32356301c68860a4ab21ed
+// https://github.com/confluentinc/rest-utils/blob/master/core/src/main/java/io/confluent/rest/FileWatcher.java
+/**
+ * Watches a file and calls a callback when it is changed.
+ */
+public class FileWatcher extends Thread {
+
+  private static final Logger log = LoggerFactory.getLogger(FileWatcher.class);
+
+  public interface Callback {
+    void run() throws Exception;
+  }
+
+  private volatile boolean shutdown;
+  private final WatchService watchService;
+  private final Path file;
+  private final Callback callback;
+
+  @SuppressFBWarnings(
+      value = "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE",
+      justification = "Null check on file.getParent() is present"
+  )
+  public FileWatcher(final Path file, final Callback callback) throws IOException {
+    this.file = Objects.requireNonNull(file);
+    Objects.requireNonNull(file.getParent(), "Watch location must have parent");
+    this.watchService = FileSystems.getDefault().newWatchService();
+    // Listen to both CREATE and MODIFY to reload, which handles delete then create.
+    file.getParent().register(watchService,
+        StandardWatchEventKinds.ENTRY_CREATE,
+        StandardWatchEventKinds.ENTRY_MODIFY);
+    this.callback = Objects.requireNonNull(callback);
+  }
+
+  /**
+   * Closes the file watcher
+   */
+  public void shutdown() {
```

Comment:
Hey @purplefox , now that `FileWatcher` extends `Thread`, do you think it makes sense to get rid of this `shutdown()` method and instead call `interrupt()` to terminate the file watcher? I'm wondering whether it's canonical to have a "clean" shutdown method like this, or if `interrupt()` is understood to be used for that purpose.

Reply:
Merging PR for now -- will make this improvement in a follow-up (if we think it'd be an improvement).

Reply:
I think it's preferable to shut cleanly without interrupting threads if you can. Interrupting threads is only necessary if the thread is blocked on something that is interruptable. In our case we only call shutdown() from the thread itself so we can guarantee it's not blocked so there's no need to interrupt anything.
If the method was intended to be used elsewhere from a different thread then we should call Thread.interrupt() but that's not the case here.
As a sanity check you could add this in shutdown() to make sure that is the case
```
if (this != Thread.currentThread()) {
   throw new IllegalStateException("Can only be called from file watcher thread");
}
```